packages/frontend-2/components/viewer/models/TreeItem.vue

```diff
@@ -444,4 +394,122 @@ const isolateOrUnisolateObject = () => {
 
   unIsolateObjects(ids)
 }
+
+const setSelection = (e: MouseEvent) => {
+  disableScrollOnNextSelection.value = true
+
+  if (isSelected.value && !e.shiftKey) {
+    // If already selected, try to expand if possible, but don't deselect
+    if ((isSingleCollection.value || isMultipleCollection.value) && !unfold.value) {
+      unfold.value = true
+      emit('expanded', props.depth)
+    }
+    disableScrollOnNextSelection.value = false
+    return
+  }
+  if (isSelected.value && e.shiftKey) {
+    removeFromSelection(rawSpeckleData.value)
+    disableScrollOnNextSelection.value = false
+    return
+  }
+  if (!e.shiftKey) clearSelection()
+  addToSelection(rawSpeckleData.value)
+
+  // Auto-expand when selecting if it has children
+  if ((isSingleCollection.value || isMultipleCollection.value) && !unfold.value) {
+    unfold.value = true
+    emit('expanded', props.depth)
+  }
+}
+
+const ensureSelectedItemsVisible = () => {
+  const selectedIds = objects.value.map((obj: { id: string }) => obj.id)
+  if (selectedIds.length === 0) return
+
+  // Check if any selected items are in this container but beyond current pagination
+  const allItems = singleCollectionItems.value
+  const selectedIndices = allItems
+    .map((item, index) => ({ item, index }))
+    .filter(({ item }) => {
+      const itemId = item.rawNode.raw?.id
+      return itemId && selectedIds.includes(itemId)
+    })
+    .map(({ index }) => index)
+
+  if (selectedIndices.length > 0) {
+    const maxSelectedIndex = Math.max(...selectedIndices)
+    // Ensure itemCount is high enough to show the selected item
+    if (maxSelectedIndex >= itemCount.value) {
+      itemCount.value = Math.max(maxSelectedIndex + 1, itemCount.value)
+    }
+  }
+}
+
+const checkAndExpand = () => {
+  const nodeId = rawSpeckleData.value.id
+  const shouldForceExpand = props.forceExpandedNodeIds?.has(nodeId)
+
+  if (
+    (isSingleCollection.value || isMultipleCollection.value) &&
+    (props.expandLevel >= props.depth || shouldForceExpand)
+  ) {
+    unfold.value = true
+    ensureSelectedItemsVisible()
+  }
+}
+
+const headerElement = ref<HTMLElement>()
+const disableScrollOnNextSelection = ref(false)
+
+onMounted(() => {
+  checkAndExpand()
+
+  const nodeId = rawSpeckleData.value.id
+  const isForceExpanded = props.forceExpandedNodeIds?.has(nodeId)
+
+  if (
+    isSelected.value &&
+    isForceExpanded &&
+    headerElement.value &&
+    !disableScrollOnNextSelection.value
+  ) {
+    headerElement.value.scrollIntoView({
+      behavior: 'instant',
+      block: 'center'
+    })
+  }
+
+  disableScrollOnNextSelection.value = false
+})
+
+watch(
+  () => props.forceExpandedNodeIds,
+  () => {
+    checkAndExpand()
+  },
+  { deep: true }
```

Comment:
why use a deep watcher if its a single dimensional array/set? `deep` should be used sparingly because of the performance implications